pkg/reconciler/trigger/controller.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trigger
+
+import (
+	"context"
+	"k8s.io/client-go/tools/cache"
+
+	"cloud.google.com/go/pubsub"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"knative.dev/eventing/pkg/apis/eventing"
+	eventingv1beta1 "knative.dev/eventing/pkg/apis/eventing/v1beta1"
+	"knative.dev/eventing/pkg/duck"
+	"knative.dev/eventing/pkg/logging"
+	"knative.dev/pkg/client/injection/ducks/duck/v1/addressable"
+	"knative.dev/pkg/client/injection/ducks/duck/v1/conditions"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	pkgcontroller "knative.dev/pkg/controller"
+	pkgreconciler "knative.dev/pkg/reconciler"
+	"knative.dev/pkg/resolver"
+
+	brokerv1beta1 "github.com/google/knative-gcp/pkg/apis/broker/v1beta1"
+	brokerinformer "github.com/google/knative-gcp/pkg/client/injection/informers/broker/v1beta1/broker"
+	triggerinformer "github.com/google/knative-gcp/pkg/client/injection/informers/broker/v1beta1/trigger"
+	triggerreconciler "github.com/google/knative-gcp/pkg/client/injection/reconciler/broker/v1beta1/trigger"
+	metadataClient "github.com/google/knative-gcp/pkg/gclient/metadata"
+	"github.com/google/knative-gcp/pkg/reconciler"
+	"github.com/google/knative-gcp/pkg/utils"
+)
+
+const (
+	// controllerAgentName is the string used by this controller to identify
+	// itself when creating events.
+	controllerAgentName = "trigger-controller"
+	// finalizerName is the name of the finalizer that this controller adds to the Triggers that it reconciles.
+	finalizerName = "googlecloud"
+)
+
+// filterBroker is the function to filter brokers with proper brokerclass.
+var filterBroker = pkgreconciler.AnnotationFilterFunc(eventingv1beta1.BrokerClassAnnotationKey, brokerv1beta1.BrokerClass, false /*allowUnset*/)
+
+func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+	// TODO initialize project ID here via env var or metadata.
```

Comment:
Do we want to support reconciling across multiple projects?

Reply:
I would say no for now unless there are some compelling use cases. Do you have any use case in mind ?

Reply:
I haven't heard of a need for reconciling across multiple projects.

Notes for a future PR: Using an env var here makes tests more complicated. One option for setting project ID in this and other controllers is to add it to the context created in cmd/controller/main.go. Then controllers have a uniform way of getting the global project ID if they need it, and there's only one piece of code responsible for retrieving the project id from metadata or env var.